packages/salesforcedx-apex-replay-debugger/src/core/heapDumpService.ts

```diff
@@ -136,49 +127,123 @@ export class HeapDumpService {
-      varContainer.variables.forEach((value, key) => {
-        for (const variableEntry of heapDumpExtentValue.entry!) {
-          const valueAsApexVar = value as ApexVariableContainer;
-          this.updateVariableContainerWithExtentValueOrEntry(
+      this.updateContainerChildrenWithEntries(
+        varContainer,
+        heapDumpExtentValue.entry,
+        extentToRevisit,
+        referenceToRevisit,
+        variableTypes
+      );
+    }
+  }
+
+  public updateContainerChildrenWithEntries(
+    varContainer: ApexVariableContainer,
+    extentValueEntries: HeapDumpExtentValueEntry[],
+    extentToRevisit: Map<string, HeapDumpExtentValue>,
+    referenceToRevisit: Map<string, ApexVariableContainer>,
+    variableTypes: Map<string, string>
+  ): void {
+    for (const extentValueEntry of extentValueEntries) {
+      let foundMatchingApexVariable = false;
+      for (const entry of Array.from(varContainer.variables.entries())) {
+        const valueAsApexVar = entry[1] as ApexVariableContainer;
+        if (this.isContainerForExtentEntry(valueAsApexVar, extentValueEntry)) {
+          this.updateContainerOrCache(
             valueAsApexVar,
-            variableEntry.value.value,
-            variableEntry.value.entry
+            extentValueEntry,
+            extentToRevisit,
+            referenceToRevisit,
+            variableTypes
           );
+          foundMatchingApexVariable = true;
+          break;
         }
-      });
+      }
+      if (!foundMatchingApexVariable) {
+        const variableType =
+          variableTypes.get(extentValueEntry.keyDisplayValue) || '';
+        varContainer.variables.set(
+          extentValueEntry.keyDisplayValue,
+          new ApexVariableContainer(
+            extentValueEntry.keyDisplayValue,
+            extentValueEntry.value.value,
+            variableType
+          )
+        );
+      }
     }
   }
 
-  public updateVariableContainerWithExtentValueOrEntry(
+  public updateContainerWithExtentValueOrEntry(
     varContainer: ApexVariableContainer,
     extentValue: any,
-    extentEntry: HeadpDumpExtentValueEntry[] | undefined
+    extentEntry: HeapDumpExtentValueEntry[] | undefined,
+    type?: string
   ): void {
     if (extentValue) {
-      this.updateVariableContainerWithExtentValue(varContainer, extentValue);
+      this.updateContainerWithExtentValue(varContainer, extentValue, type);
     } else if (extentEntry) {
       for (const extentValueEntry of extentEntry) {
-        if (extentValueEntry.keyDisplayValue === varContainer.name) {
-          return this.updateVariableContainerWithExtentValueOrEntry(
+        if (this.isContainerForExtentEntry(varContainer, extentValueEntry)) {
+          return this.updateContainerWithExtentValueOrEntry(
             varContainer,
             extentValueEntry.value.value,
             extentValueEntry.value.entry
           );
         }
         varContainer.variables.forEach((value, key) => {
```

Comment:
This inner loop isn't correct. What this will end up doing is that it'll end up assigning each variable in the container to the extent values that it's looping through in the outer loop. When it's done, every variable will end up having the value of the last extent. It should only be updating the value for the key that matches the keyDisplayValue in the HeapDumpExtentValueEntry

Reply:
Did you fix that in your current work iteM? The other day you mentioned fixing a couple of existing bugs. One was `if (heapDumpExtentValue.value) {` is with a boolean value. What's the other one?

Reply:
@JimSuplizio ^^^

Reply:
@DatGuyJonathan At this point I've got the fixes locally. If this is working for you then check it in and I'll update for the loop and value checks in my version of your branch.